File: Data Structures & Algorithms/course-schedule/submission-8.java
class Solution {
    Map<Integer,List<Integer>> adj = new HashMap<>();
    Map<Integer,Boolean> vis = new HashMap<>();
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        for(int i = 0;i<numCourses;i++){
            adj.put(i,new ArrayList<>());
        }
        for(int[] pre:prerequisites){
            int crs=pre[0], p = pre[1];
            adj.get(crs).add(p);
        }
        for(int i= 0; i<numCourses;i++){
            if(!dfs(i))
                return false;
        }
        return true;
    }

    public boolean dfs(int crs) {
        if(vis.containsKey(crs)){
            return vis.get(crs);
        }
        vis.put(crs,false);
        for(int pre:adj.get(crs)){
            if(!dfs(pre))
                return false;
        }
        vis.put(crs,true);
        return true;
    }
}

File: Data Structures & Algorithms/count-connected-components/submission-8.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:

        vis = set()
        adj = {i:[] for i in range(n)}
        def dfs(n):
            if n in vis:
                return
            vis.add(n)
            for nei in adj[n]:
                dfs(nei)
        for n1,n2 in edges:
            adj[n1].append(n2)
            adj[n2].append(n1)
        res = 0
        for i in range(n):
            if i not in vis:
                res+=1
                dfs(i)
        return res

        